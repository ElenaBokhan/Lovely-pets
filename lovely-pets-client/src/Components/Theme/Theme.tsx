import {createContext, useEffect, useState} from 'react';
import {THEME_STORAGE_KEY} from 'Const';

const supportedThemes = {
    light: 'light',
    dark: 'dark',
};

type Themes = keyof typeof supportedThemes;

export const ThemeContext = createContext<
    | {
          theme: Themes;
          setTheme: (theme: Themes) => void;
          supportedThemes: {[key: string]: string};
      }
    | undefined
>(undefined);

const getTheme = (): Themes => {
    let theme = localStorage.getItem(THEME_STORAGE_KEY);

    if (!theme) {
        localStorage.setItem(THEME_STORAGE_KEY, 'light');
        theme = 'light';
    }

    return theme as Themes;
};

const Theme = (props: {children: React.ReactNode}) => {
    const [theme, setTheme] = useState<Themes>(getTheme);

    useEffect(() => {
        localStorage.setItem(THEME_STORAGE_KEY, theme);
        document.documentElement.setAttribute('data-theme', theme);
    }, [theme]);

    return (
        <ThemeContext.Provider
            value={{
                theme,
                setTheme,
                supportedThemes,
            }}
        >
            {props.children}
        </ThemeContext.Provider>
    );
};

export default Theme;
