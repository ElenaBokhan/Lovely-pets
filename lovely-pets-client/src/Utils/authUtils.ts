import {ILoginResponse} from 'Api/authClient';

export const setLoginToStorage = (loginData: ILoginResponse) => {
    const {access_token, refresh_token, username, id} = loginData;

    localStorage.setItem('access_token', access_token);
    localStorage.setItem('refresh_token', refresh_token);
    localStorage.setItem('userName', username);
    localStorage.setItem('userId', id);
};

export const removeLoginToStorage = () => {
    localStorage.removeItem('access_token');
    localStorage.removeItem('refresh_token');
    localStorage.removeItem('userName');
    localStorage.removeItem('userId');
};
