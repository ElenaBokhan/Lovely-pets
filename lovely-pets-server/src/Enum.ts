export enum ESortFilter {
    POPULAR = 'POPULAR',
    NEW = 'NEW',
    CHEAPER = 'CHEAPER',
    EXPENSIVE = 'EXPENSIVE',
    RATING = 'RATING',
    DISCOUNT = 'DISCOUNT',
}
