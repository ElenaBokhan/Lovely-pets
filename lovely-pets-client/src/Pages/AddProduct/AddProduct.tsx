import {Button} from 'Components/Common/Button/Button';
import {TitlePage} from 'Components/Common/TitlePage/TitlePage';
import styles from 'Pages/AddReview/AddReview.module.css';
import {useState} from 'react';
import {Gap} from 'Components/Common/Gap/Gap';
import {FieldForm} from 'Components/Common/FieldForm/FieldForm';
import {useNavigate} from 'react-router-dom';
import {ETestId} from 'Enum';
import {useCreatePetMutation} from 'Api/queryHooks';

export interface IPetField {
    label: string;
    value: string | number;
    param: keyof INewProduct;
    testId: ETestId;
}

export const AddProduct = () => {
    const {mutateAsync: createPet} = useCreatePetMutation();
    const navigate = useNavigate();

    const getInitialProduct = (): IPetField[] => [
        {
            param: 'name',
            label: 'Название товара',
            value: '',
            testId: ETestId.ADD_PRODUCT_NAME_INPUT,
        },
        {
            param: 'price',
            label: 'Цена товара',
            value: '',
            testId: ETestId.ADD_PRODUCT_PRICE_INPUT,
        },
        {
            param: 'description',
            label: 'Описание',
            value: '',
            testId: ETestId.ADD_PRODUCT_DESCRIPTION_INPUT,
        },
        {
            param: 'discount',
            label: 'Скидка',
            value: '',
            testId: ETestId.ADD_PRODUCT_DISCOUNT_INPUT,
        },
        {
            param: 'stock',
            label: 'Наличие',
            value: '',
            testId: ETestId.ADD_PRODUCT_STOCK_INPUT,
        },
        {
            param: 'wight',
            label: 'Вес',
            value: '',
            testId: ETestId.ADD_PRODUCT_WIGHT_INPUT,
        },
        {
            param: 'pictures',
            label: 'Добавьте фото',
            value: '',
            testId: ETestId.ADD_PRODUCT_PICTURES_INPUT,
        },
    ];

    const [pet, setPet] = useState<IPetField[]>(getInitialProduct());

    const isNotEmptyValues = (): boolean => {
        for (const field of pet) {
            if (field.value === '') return false;
        }

        return true;
    };

    const getNewProduct = (): INewProduct => {
        const newPet = {} as INewProduct;

        pet.forEach(({value, param}) => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            newPet[param] = value;
        });

        return newPet;
    };

    const handleSubmit = async (event: React.SyntheticEvent<HTMLFormElement>) => {
        event.preventDefault();

        try {
            if (isNotEmptyValues()) {
                const createdPet = await createPet(getNewProduct());

                navigate(`/${createdPet._id}`);
            }
        } catch (error) {
            console.error(error);
        }
    };

    const handleSetProduct = (value: string, param: keyof IPet) => {
        const currentValue = param === 'price' || param === 'discount' || param === 'stock' ? Number(value) : value;

        const newProducts: IPetField[] = pet.map((item) =>
            item.param === param ? {...item, value: currentValue} : item
        );
        setPet(newProducts);
    };

    const renderAddProductForm = () => {
        return (
            <form onSubmit={handleSubmit}>
                {pet.map((field) => (
                    <FieldForm key={field.label} onChange={handleSetProduct} field={field} />
                ))}

                <Button testId={ETestId.ADD_PRODUCT_SUBMIT_BUTTON} type="submit" label={'Добавить котика'} />
            </form>
        );
    };

    return (
        <div className={styles.addReviewPage}>
            <TitlePage label={'Добавление нового товара'} />
            <Gap size={40} />
            <hr />
            {renderAddProductForm()}
        </div>
    );
};
