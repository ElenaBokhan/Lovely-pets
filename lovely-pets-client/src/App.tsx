import './index.css';
import {Layout} from 'Components/Layout/Layout';
import {Catalog} from 'Pages/Catalog/Catalog';
import {Favourites} from 'Pages/Favourites/Favourites';
import {PageNotFound} from 'Pages/PageNotFound/PageNotFound';
import {PetCard} from 'Pages/PetCard/PetCard';
import {Profile} from 'Pages/Profile/Profile';
import {RouterProvider, createBrowserRouter, RouteObject} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';
import {SignUp} from 'Pages/SignUp/SignUp';
import {SignIn} from 'Pages/SignIn/SignIn';
import {Cart} from 'Pages/Cart/Cart';
import {AddReview} from 'Pages/AddReview/AddReview';
import {AddProduct} from 'Pages/AddProduct/AddProduct';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import {withTranslation} from 'react-i18next';
import {useEffect} from 'react';
import {setCartFromStorage, setCartToStorage} from 'Store/cartStore';
import 'Style/Themes.module.css';
import Theme from 'Components/Theme/Theme';

export const router: RouteObject[] = [
    {
        path: '/',
        element: <Layout />,
        errorElement: <div />, // TODO: добавить errorBoundary, пока заглушка
        children: [
            {
                path: '/',
                element: <Catalog />,
            },
            {
                path: '/favourites',
                element: <Favourites />,
            },
            {
                path: '/pets/:petId',
                element: <PetCard />,
            },
            {
                path: '/profile',
                element: <Profile />,
            },
            {
                path: '/addReview',
                element: <AddReview />,
            },
            {
                path: '/addPet',
                element: <AddProduct />,
            },
            {
                path: '/signup',
                element: <SignUp />,
            },
            {
                path: '/signin',
                element: <SignIn />,
            },
            {
                path: '/cart',
                element: <Cart />,
            },
            {
                path: '*',
                element: <PageNotFound />,
            },
        ],
    },
];

export const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            retry: false,
        },
    },
});

const App = () => {
    useEffect(() => {
        window.addEventListener('load', setCartFromStorage);
        window.addEventListener('beforeunload', setCartToStorage);
        return () => {
            window.removeEventListener('load', setCartFromStorage);
            window.removeEventListener('beforeunload', setCartToStorage);
        };
    }, []);

    return (
        <QueryClientProvider client={queryClient}>
            <Theme>
                <RouterProvider router={createBrowserRouter(router)} />
                <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
            </Theme>
        </QueryClientProvider>
    );
};

export default withTranslation()(App);
