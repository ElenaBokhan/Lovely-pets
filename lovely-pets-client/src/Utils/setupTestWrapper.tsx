import {queryClient, router} from 'App';
import {render} from '@testing-library/react';
import {createMemoryRouter, RouterProvider} from 'react-router-dom';
import {ToastContainer} from 'react-toastify';
import {QueryClientProvider} from '@tanstack/react-query';
import Theme from 'Components/Theme/Theme';

interface WrapperForTestsParams {
    initialRouterEntries: string[];
}

export const setupTestsWrapper = ({initialRouterEntries}: WrapperForTestsParams) => {
    const memoryRouter = createMemoryRouter(router, {
        initialEntries: initialRouterEntries,
    });
    render(
        <QueryClientProvider client={queryClient}>
            <Theme>
                <RouterProvider router={memoryRouter} />
                <ToastContainer autoClose={2000} position="top-right" theme="light" closeOnClick />
            </Theme>
        </QueryClientProvider>
    );

    return memoryRouter;
};
