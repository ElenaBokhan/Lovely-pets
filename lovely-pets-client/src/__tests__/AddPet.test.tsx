import '@testing-library/jest-dom';
import {setupTestsWrapper} from 'Utils/setupTestWrapper';
import {fireEvent, render, screen} from '@testing-library/react';
import {BrowserRouter, createMemoryRouter} from 'react-router-dom';
import {ETestId} from 'Enum';
import {userEvent} from '@testing-library/user-event';
import {AddProduct} from 'Pages/AddProduct/AddProduct';
import {QueryClientProvider} from '@tanstack/react-query';
import Theme from 'Components/Theme/Theme';
import {queryClient} from 'App';

describe('Тесты для компонента addProduct', () => {
    let router: ReturnType<typeof createMemoryRouter> | undefined;
    let submitBtn: HTMLButtonElement | undefined;
    let nameInput: HTMLInputElement | undefined;
    let priceInput: HTMLInputElement | undefined;
    let descriptionInput: HTMLInputElement | undefined;
    let discountInput: HTMLInputElement | undefined;
    let stockInput: HTMLInputElement | undefined;
    let wightInput: HTMLInputElement | undefined;
    let picturesInput: HTMLInputElement | undefined;
    let allInputs: HTMLInputElement[] | undefined;

    beforeEach(() => {
        router = setupTestsWrapper({initialRouterEntries: ['/addPet']});
        submitBtn = screen.getByTestId(ETestId.ADD_PRODUCT_SUBMIT_BUTTON);
        nameInput = screen.getByTestId(ETestId.ADD_PRODUCT_NAME_INPUT);
        priceInput = screen.getByTestId(ETestId.ADD_PRODUCT_PRICE_INPUT);
        descriptionInput = screen.getByTestId(ETestId.ADD_PRODUCT_DESCRIPTION_INPUT);
        discountInput = screen.getByTestId(ETestId.ADD_PRODUCT_DISCOUNT_INPUT);
        stockInput = screen.getByTestId(ETestId.ADD_PRODUCT_STOCK_INPUT);
        wightInput = screen.getByTestId(ETestId.ADD_PRODUCT_WIGHT_INPUT);
        picturesInput = screen.getByTestId(ETestId.ADD_PRODUCT_PICTURES_INPUT);
        allInputs = [nameInput, priceInput, descriptionInput, discountInput, stockInput, wightInput, picturesInput];
    });

    describe('Проверяем доступность элементов компонента AddPet', () => {
        test('Проверяем что все поля доступны', () => {
            checkInputsEnabled(allInputs);
        });

        test('Проверяем что кнопка добавления продукта доступна', () => {
            expect(submitBtn).toBeEnabled();
        });
    });

    describe('Проверяем работу компонента AddProduct при заполненных и незаполлненных полях', () => {
        test('Проверяем что запрос не выполняется при незаполненных полях и редиректа не происходит', async () => {
            checkInputsHasEmptyValue(allInputs);

            await userEvent.click(submitBtn);
            expect(router?.state.location.pathname).toEqual('/addPet');
        });

        test('Проверяем что поля заполняются верно', async () => {
            const stringInputValue = 'Значение';
            const numberInputValue = 123;

            const stringInputs = [nameInput, descriptionInput, picturesInput, wightInput];
            const numberInputs = [priceInput, stockInput, discountInput];

            fillInputsWithValue(stringInputs, stringInputValue);
            fillInputsWithValue(numberInputs, numberInputValue);

            checkFillInputsValue(stringInputs, stringInputValue);
            checkFillInputsValue(numberInputs, numberInputValue);
        });
    });

    describe('Снапшот компонента AddProduct', () => {
        test('Snapshot', () => {
            const {asFragment} = render(
                <QueryClientProvider client={queryClient}>
                    <Theme>
                        <BrowserRouter>
                            <AddProduct />
                        </BrowserRouter>
                    </Theme>
                </QueryClientProvider>
            );

            expect(asFragment()).toMatchSnapshot();
        });
    });
});

function checkInputsEnabled(allInputs: HTMLInputElement[]) {
    allInputs.forEach((input) => {
        expect(input).toBeEnabled();
    });
}

function checkInputsHasEmptyValue(allInputs: HTMLInputElement[]) {
    allInputs.forEach((input) => {
        expect(input).toHaveDisplayValue('');
    });
}

async function fillInputsWithValue(allInputs: HTMLInputElement[], value: string | number) {
    allInputs.forEach(async (input) => {
        await fireEvent.change(input, {target: {value}});
    });
}

function checkFillInputsValue(allInputs: HTMLInputElement[], value: string | number) {
    allInputs.forEach(async (input) => {
        expect(input).toHaveDisplayValue(value.toString());
    });
}
