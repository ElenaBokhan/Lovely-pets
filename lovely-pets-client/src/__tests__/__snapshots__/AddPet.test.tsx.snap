// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тесты для компонента addProduct Снапшот компонента AddProduct Snapshot 1`] = `
<DocumentFragment>
  <div
    class="addReviewPage"
  >
    <div
      data-testid="titlePage"
    >
      <a
        class="breadcrumbs"
        data-testid="breadcrumbsButton"
        href="/"
      >
        <img
          alt="arrow"
          src="SvgrURL"
        />
        <span
          class="grey undefined subscribe2"
        >
          Назад
        </span>
      </a>
      <span
        class="undefined undefined head1"
      >
        Добавление нового товара
      </span>
    </div>
    <div
      style="height: 40px;"
    />
    <hr />
    <form>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Название товара
        </span>
        <input
          class="fieldTextarea"
          data-test-id="addProductNameInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Цена товара
        </span>
        <input
          class="fieldTextarea"
          data-test-id="addProductPriceInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Описание
        </span>
        <input
          class="fieldTextarea"
          data-test-id="addProductDescriptionInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Скидка
        </span>
        <input
          class="fieldTextarea"
          data-test-id="addProductDiscountInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Наличие
        </span>
        <input
          class="fieldTextarea"
          data-test-id="addProductStockInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Вес
        </span>
        <input
          class="fieldTextarea"
          data-test-id="addProductWightInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <span
          class="undefined undefined paragraph1"
        >
          Добавьте фото
        </span>
        <input
          class="fieldTextarea"
          data-test-id="addProductPicturesInput"
          placeholder="Укажите значение"
          type="text"
          value=""
        />
      </div>
      <button
        class="button standard"
        data-test-id="addProductSubmitButton"
        data-value="Добавить котика"
        type="submit"
      >
        Добавить котика
      </button>
    </form>
  </div>
</DocumentFragment>
`;
